src/Common/ThreadStatus.h

```diff
@@ -57,45 +57,57 @@ using ThreadStatusPtr = ThreadStatus *;
   * Create via CurrentThread::initializeQuery (for queries) or directly (for various background tasks).
   * Use via CurrentThread::getGroup.
   */
+class ThreadGroupStatus;
+using ThreadGroupStatusPtr = std::shared_ptr<ThreadGroupStatus>;
+
 class ThreadGroupStatus
 {
 public:
-    struct ProfileEventsCountersAndMemory
-    {
-        ProfileEvents::Counters::Snapshot counters;
-        Int64 memory_usage;
-        UInt64 thread_id;
-    };
-
-    mutable std::mutex mutex;
+    /// The first thread created this thread group
+    UInt64 master_thread_id = 0;
 
     ProfileEvents::Counters performance_counters{VariableContext::Process};
     MemoryTracker memory_tracker{VariableContext::Process};
 
-    ContextWeakPtr query_context;
-    ContextWeakPtr global_context;
+    /// Access to the members below has to be in critical section with  mutex
+    mutable std::mutex mutex;
 
     InternalTextLogsQueueWeakPtr logs_queue_ptr;
     InternalProfileEventsQueueWeakPtr profile_queue_ptr;
```

Comment:
So maybe add getters/setters for these fields as well? (and move to the private part).
Also, it's interesting that `logs_queue_ptr` and `client_logs_level` are actually logically connected (initialized in the same place, and are needed for the same feature - writing to text log). It's interesting why they were so separated here...